src/nnvm/low_precision_pass.cc

```diff
@@ -0,0 +1,265 @@
+      "conditional_fp32_ops");
+
+  CHECK(target_dtype == mshadow::kFloat16)
+      << "Only float16 target_dtype is supported yet";
+
+  // Additional data structures to share common cast node inputs among different nodes
+  std::unordered_map<Node *, NodePtr> mirror_map;
+  nnvm::NodeEntryMap<NodeEntry> mirror_fp32_map;
+  nnvm::NodeEntryMap<NodeEntry> mirror_target_dtype_map;
+
+  // Visit nodes in a topologically sorted order
+  DFSVisit(src.outputs, [&](const NodePtr &node) {
+    NodePtr new_node = Node::Create();
+    *new_node = *node;
+    new_node->inputs.clear();
+
+    /* 1. for node which needs to run in FP32 mode, add amp_cast operators
+     * (to fp32) after its inputs
+     * 2. for node which needs to run in FP16 mode, add amp_cast operators
+     * (to target_dtype) after its inputs
+     * 3. for nodes which need to run in widest dtype among its inputs, add
+     * amp_multicast operators between op and its inputs
+     * 4. for nodes which need to run in FP32 mode, based on a specific condition,
+     * check the condition, and if true add amp_cast (to fp32) after its inputs
+     * 4. for other nodes, create copy node and add it to the mirror_map
+     */
+    if (!node->is_variable() && fp32_ops.count(node->op()->name) > 0 &&
+        excluded_syms.count(node->attrs.name) == 0) {
+      for (size_t i = 0; i < node->inputs.size(); ++i) {
+        const auto &e = node->inputs[i];
+        if (mirror_fp32_map.count(e)) {
+          new_node->inputs.emplace_back(mirror_fp32_map[e]);
+        } else {
+          NodePtr mirror_node = mirror_map.at(e.node.get());
+          NodeEntry mirror_entry = NodeEntry{mirror_node, e.index, e.version};
+          std::string suffix = GetSuffix(e, mirror_map);
+          AddCastNode(e, suffix, mirror_entry, "float32", &mirror_fp32_map,
+                      new_node);
+        }
+      }
+    } else if (!node->is_variable() &&
+               target_dtype_ops.count(node->op()->name) > 0 &&
+               excluded_syms.count(node->attrs.name) == 0) {
+      for (size_t i = 0; i < node->inputs.size(); ++i) {
+        const auto &e = node->inputs[i];
+        if (mirror_target_dtype_map.count(e)) {
+          new_node->inputs.emplace_back(mirror_target_dtype_map[e]);
+        } else {
+          NodePtr mirror_node = mirror_map.at(e.node.get());
+          NodeEntry mirror_entry = NodeEntry{mirror_node, e.index, e.version};
+          std::string suffix = GetSuffix(e, mirror_map);
+          AddCastNode(e, suffix, mirror_entry, "float16",
+                      &mirror_target_dtype_map, new_node);
+        }
+      }
+    } else if (!node->is_variable() &&
+               widest_dtype_ops.count(node->op()->name) > 0 &&
+               excluded_syms.count(node->attrs.name) == 0) {
+      CHECK(node->inputs.size() > 0)
+          << "Please check the symbol. node name: " << node->attrs.name
+          << "op name " << node->op()->name << " has no inputs";
```

Comment:
If this check fails, what does it mean? And what should the user do if it fails? The message says "check the name" but what are they supposed to check for, and why did this fail?

Can you improve the error message and add some more comments to explain to devs what they should do if this check fails?

Reply:
They should check the node name to see that the input is part of the symbol or not. If not maybe something went wrong with symbol construction. I can add more comment to check if something went wrong with symbol construction, but not much more information I can provide here.